qiskit/primitives/containers/bindings_array.py

```diff
@@ -0,0 +1,355 @@
+                arrays over the last axis is equivalent to providing them separately.
+            kwvals: A mapping from one or more parameters to arrays of values to bind
+                them to, where the last axis is over parameters.
+            shape: The leading shape of every array in these bindings.
+
+        Raises:
+            ValueError: If all inputs are ``None``.
+            ValueError: If the shape cannot be automatically inferred from the arrays, or if there
+                is some inconsistency in the shape of the given arrays.
+        """
+        super().__init__()
+
+        if vals is None:
+            vals = []
+        if kwvals is None:
+            kwvals = {}
+
+        vals = [vals] if isinstance(vals, np.ndarray) else [np.array(v, copy=False) for v in vals]
+        # TODO str will be used for internal data (_kwvals) instead of Parameter.
+        # This requires https://github.com/Qiskit/qiskit/issues/7107
+        kwvals = {
+            (p,) if isinstance(p, Parameter) else tuple(p): np.array(val, copy=False)
+            for p, val in kwvals.items()
+        }
+
+        if shape is None:
+            # jump through hoops to find out user's intended shape
+            shape = _infer_shape(vals, kwvals)
+
+        # shape checking, and normalization so that each last index must be over parameters
+        self._shape = shape_tuple(shape)
+        for idx, val in enumerate(vals):
+            vals[idx] = _standardize_shape(val, self._shape)
+
+        self._vals: list[np.ndarray] = vals
+        self._kwvals = kwvals
+
+        self.validate()
+
+    def __getitem__(self, args) -> BindingsArray:
+        # because the parameters live on the last axis, we don't need to do anything special to
+        # accomodate them because there will always be an implicit slice(None, None, None)
+        # on all unspecified trailing dimensions
+        # separately, we choose to not disallow args which touch the last dimension, even though it
+        # would not be a particularly friendly way to chop parameters
+        vals = [val[args] for val in self._vals]
+        kwvals = {params: val[args] for params, val in self._kwvals.items()}
+        try:
+            shape = next(chain(vals, kwvals.values())).shape[:-1]
+        except StopIteration:
+            shape = ()
+        return BindingsArray(vals, kwvals, shape)
+
+    @property
+    def kwvals(self) -> Dict[Tuple[str, ...], np.ndarray]:
+        """The keyword values of this array."""
+        return {_format_key(k): v for k, v in self._kwvals.items()}
+
+    @property
+    def num_parameters(self) -> int:
+        """The total number of parameters."""
+        return sum(val.shape[-1] for val in chain(self.vals, self._kwvals.values()))
+
+    @property
+    def vals(self) -> List[np.ndarray]:
+        """The non-keyword values of this array."""
+        return self._vals
+
+    def bind(self, circuit: QuantumCircuit, loc: Tuple[int, ...]) -> QuantumCircuit:
```

Comment:
```suggestion
    def bind(self, circuit: QuantumCircuit, index: Tuple[int, ...]) -> QuantumCircuit:
```
Personally i prefer index to loc. When talking about ndarrays you talk about the index, not the loc. Eg `np.ndindex(shape)`.

Reply:
This point was corrected in @ihincks's comment, so please talk about which you prefer.